c2rust-analyze/src/context.rs

```diff
@@ -445,11 +459,51 @@ impl<'tcx> TypeOf<'tcx> for PlaceRef<'tcx> {
     }
 }
 
+fn const_alloc_id(c: &Constant) -> Option<AllocId> {
+    if let ConstantKind::Val(ConstValue::Scalar(interpret::Scalar::Ptr(ptr, _)), _ty) = c.literal {
+        return Some(ptr.provenance);
+    }
+    None
+}
+
+fn find_static_for_alloc<'tcx>(tcx: &TyCtxt<'tcx>, id: AllocId) -> Option<DefId> {
+    match tcx.try_get_global_alloc(id) {
+        None => {} //hmm, ok
+        Some(GlobalAlloc::Static(did)) => {
+            if !tcx.is_foreign_item(did) {
+                // local static referenced
+                return Some(did);
+            }
+        }
+        Some(_) => {}
+    }
+    None
+}
+
 impl<'tcx> TypeOf<'tcx> for Operand<'tcx> {
     fn type_of(&self, acx: &AnalysisCtxt<'_, 'tcx>) -> LTy<'tcx> {
         match *self {
             Operand::Move(pl) | Operand::Copy(pl) => acx.type_of(pl),
-            Operand::Constant(ref c) => label_no_pointers(acx, c.ty()),
+            Operand::Constant(ref c) => {
+                // Constants of pointer type should only be pointers into static allocations.
+                // Find the defid of the static and look up the pointer ID in gacx.static_tys
+                if c.ty().is_any_ptr() {
+                    if let Some(alloc_id) = const_alloc_id(c) {
+                        if let Some(did) = find_static_for_alloc(&acx.gacx.tcx, alloc_id) {
+                            match acx.gacx.static_tys.get(&did) {
+                                Some(lty) => lty,
+                                None => panic!("did {:?} not found", did),
+                            }
+                        } else {
+                            panic!("no static found for alloc id {:?}", alloc_id)
+                        }
+                    } else {
+                        panic!("constant was of ptr type but not a Scalar pointing into a static allocation: {:?}", c)
+                    }
```

Comment:
nit: I think these `if let Some(...) = ... { ... } else { panic!(...); }`s can be `.unwrap_or_else(|| panic!(...))`s and that would get rid of the nesting.